homeassistant/components/reolink/host.py

```diff
@@ -363,32 +364,51 @@ async def handle_webhook(
         self, hass: HomeAssistant, webhook_id: str, request: Request
     ):
         """Shield the incoming webhook callback from cancellation."""
-        await asyncio.shield(self.handle_webhook_shielded(hass, webhook_id, request))
+        self._webhook_read_done.clear()
+        shielded_task = hass.async_create_task(
+            self._handle_webhook(hass, webhook_id, request)
+        )
+        asyncio.shield(shielded_task)
+        _LOGGER.debug("Webhook '%s' called", webhook_id)
+        await self._webhook_read_done.wait()
 
-    async def handle_webhook_shielded(
+    async def _handle_webhook(
         self, hass: HomeAssistant, webhook_id: str, request: Request
     ):
         """Handle incoming webhook from Reolink for inbound messages and calls."""
+        try:
+            try:
+                data = await request.text()
+            except ConnectionResetError:
+                _LOGGER.debug(
+                    "Webhook '%s' called, but lost connection before reading message, issuing poll",
+                    webhook_id,
+                )
+                if await self._api.get_motion_state_all_ch():
+                    async_dispatcher_send(hass, f"{webhook_id}_all", {})
+                    return
+                _LOGGER.error(
+                    "Could not poll motion state after losing connection during receiving ONVIF event"
+                )
+                return
+            finally:
+                self._webhook_read_done.set()
+                if not self._webhook_reachable.is_set():
+                    self._webhook_reachable.set()
+                    ir.async_delete_issue(self._hass, DOMAIN, "webhook_url")
 
-        _LOGGER.debug("Webhook '%s' called", webhook_id)
-        if not self._webhook_reachable.is_set():
-            self._webhook_reachable.set()
-
-        if not request.body_exists:
-            _LOGGER.debug("Webhook '%s' triggered without payload", webhook_id)
-            return
-
-        data = await request.text()
-        if not data:
-            _LOGGER.debug(
-                "Webhook '%s' triggered with unknown payload: %s", webhook_id, data
-            )
-            return
+            if not data:
+                _LOGGER.debug(
+                    "Webhook '%s' triggered with unknown payload: %s", webhook_id, data
+                )
+                return
 
-        channels = await self._api.ONVIF_event_callback(data)
+            if (channels := await self._api.ONVIF_event_callback(data)) is None:
+                async_dispatcher_send(hass, f"{webhook_id}_all", {})
+                return
 
-        if channels is None:
-            async_dispatcher_send(hass, f"{webhook_id}_all", {})
-        else:
             for channel in channels:
                 async_dispatcher_send(hass, f"{webhook_id}_{channel}", {})
+        except Exception as err:  # pylint: disable=broad-except
+            # In case handle_webhook was cancelled before the exception occurred
```

Comment:
This comment is confusing.  How can `_handle_webhook` get canceled as it shielded?

Reply:
It states `handle_webhook` not `_handle_webhook`, I mean the function above.
Will try to clearify the comment

Reply:
I clarified the comment a bit, is it clear now?

Reply:
Once you run it in a separate task its not run from the function that created the task anymore. The event loop is now going to be handling the scheduling of the function. An cancellation exception in `handle_webhook` isn't going propagate down to `_handle_webhook` unless `handle_webhook` is awaiting `_handle_webhook`, and than since its shielded it won't.  Also `asyncio.CancelledError` inherits from `BaseException` not `Exception` so this exception catch would not catch it.

Reply:
You misunderstand the purpose.
It is for when a unexpected error occurs in `_handle_webhook`, normally it would be propaged upwards to `handle_webhook` (if I understand correctly) but since their is no await on the task, the error will not be logged but lost (If I understand correctly).

This Try except is purely to log such unexpected errors like a `KeyError` or something.

Schould i remove it?

Reply:
Eleborated the comment further, is it now clear?
Or schould I just remove the whole try except?

Reply:
Thanks for explaining the cancellation, I am understanding it better every day :)